pytradfri/color.py

```diff
@@ -1,48 +1,73 @@
 from .const import (ATTR_LIGHT_COLOR_X as X, ATTR_LIGHT_COLOR_Y as Y)
 
 
-KNOWN_XY = {
-  2200: {X: 33135, Y: 27211},
-  2700: {X: 30140, Y: 26909},
-  4000: {X: 24930, Y: 24694},
-}
-KNOWN_KELVIN = KNOWN_XY.keys()
-KNOWN_X = [v[X] for v in KNOWN_XY.values()]
-MIN_KELVIN = min(KNOWN_KELVIN)
-MAX_KELVIN = max(KNOWN_KELVIN)
-MIN_X = min(KNOWN_X)
-MAX_X = max(KNOWN_X)
+MIN_KELVIN = 1667
+MAX_KELVIN = 25000
 
 
 def can_kelvin_to_xy(k):
     return MIN_KELVIN <= k <= MAX_KELVIN
 
 
-def can_x_to_kelvin(x):
-    return MIN_X <= x <= MAX_X
-
-
-def kelvin_to_xy(k):
-    if k in KNOWN_XY:
-        return KNOWN_XY[k]
-    lower_k = max(kk for kk in KNOWN_KELVIN if kk < k)
-    higher_k = min(kk for kk in KNOWN_KELVIN if kk > k)
-    offset = (k - lower_k) / (higher_k - lower_k)
-    lower = KNOWN_XY[lower_k]
-    higher = KNOWN_XY[higher_k]
-    return {
-        coord: int(offset * higher[coord] + (1 - offset) * lower[coord])
-        for coord in [X, Y]
-    }
-
-
-def x_to_kelvin(x):
-    known = next((k for k, v in KNOWN_XY.items() if v[X] == x), None)
-    if known is not None:
-        return known
-    lower_x = max(k for k in KNOWN_X if k < x)
-    higher_x = min(k for k in KNOWN_X if k > x)
-    lower = x_to_kelvin(lower_x)
-    higher = x_to_kelvin(higher_x)
-    offset = (x - lower_x) / (higher_x - lower_x)
-    return int(offset * higher + (1 - offset) * lower)
+# Scaling to 65535 range and rounding
+def xy2tradfri(x, y):
+    return (int(x*65535+0.5), int(y*65535+0.5))
+
+
+def kelvin_to_xyY(T):
+    # Sources: "Design of Advanced Color - Temperature Control System
+    #           for HDTV Applications" [Lee, Cho, Kim]
+    # and https://en.wikipedia.org/wiki/Planckian_locus#Approximation
+    # and http://fcam.garage.maemo.org/apiDocs/_color_8cpp_source.html
+    if not (1667 <= T <= 25000):
+        return None
+
+    if T <= 4000:
+        # One number differs on Wikipedia and the paper:
+        #     0.2343589 is 0.2343580 on Wikipedia... don't know why
+        x = -0.2661239*(10**9)/T**3 - 0.2343589*(10**6)/T**2 \
+            + 0.8776956*(10**3)/T + 0.17991
+    elif T <= 25000:
+        x = -3.0258469*(10**9)/T**3 + 2.1070379*(10**6)/T**2 \
+            + 0.2226347*(10**3)/T + 0.24039
+
+    if T <= 2222:
+        y = -1.1063814*x**3 - 1.3481102*x**2 + 2.18555832*x - 0.20219683
+    elif T <= 4000:
+        y = -0.9549476*x**3 - 1.37418593*x**2 + 2.09137015*x - 0.16748867
+    elif T <= 25000:
+        y = 3.081758*x**3 - 5.8733867*x**2 + 3.75112997*x - 0.37001483
+
+    x, y = xy2tradfri(x, y)
+    return {X: x, Y: y}
+
+
+def xyY_to_kelvin(x, y):
+    # This is an approximation, for information, see the source.
+    # Source: https://en.wikipedia.org/wiki/Color_temperature#Approximation
+    # Input range for x and y is 0-65535
+    n = (x/65535-0.3320) / (y/65535-0.1858)
+    kelvin = int((-449*n**3 + 3525*n**2 - 6823.3*n + 5520.33) + 0.5)
+    return kelvin if MIN_KELVIN <= kelvin <= MAX_KELVIN else None
```

Comment:
Why would we check for MIN and MAX KELVIN here ? it's that Kelvin right?

Reply:
This reverse function for xyY to Kelvin is an approximation (https://en.wikipedia.org/wiki/Color_temperature#Approximation) which is not 100% accurate.
I just didn't want to return kelvin values for which the kelvin to xyY function wouldn't work (because that one raises when a values outside of the range is passed), this is why I only wanted to return the kelvins when they are in this range.

Reply:
It's not up to this function to decide if it's a valid Kelvin, this function should only offer an approximation. The device class should check it with `can_kelvin_to_xy` to decide to return `None` or not.

Reply:
This is the reverse function, so `can_kelvin_to_xy` is not applicable, the function that would be applicable is `can_xy_to_kelvin`, which doesn't exist at the moment because I thought that by returning None if the conversion shouldn't succeed I could conserve this function.
Should I create `can_xy_to_kelvin`?

Reply:
No, you can use `can_kelvin_to_xy` to avoid returning Kelvin inside the Ikea device. However, for color utility functions, we should not be concerned with what range IKEA supports, but instead just do our best to give an approximation.

Reply:
Included what I think you meant in commit 768c8eb :)